### PR TITLE
WD-5543- Combine panel css

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Remove duplicate panel css.
- Combine features in the Panel component.

## QA

- Check that each panel works as it did before (there will be some small visual changes but the functionality should be unchanged).
- Check that the panels look OK at mobile and desktop sizes.

## Details

https://warthogs.atlassian.net/browse/WD-5543
Fixes: https://github.com/canonical/juju-dashboard/issues/1526.
Fixes: https://github.com/canonical/juju-dashboard/issues/1523.

### CODE REVIEW

File: src/components/ModelTableList/StatusGroup.tsx [hunk 1]
```diff
@@ -55,7 +55,7 @@ const generateWarningMessage = (model: ModelData) => {
       {messages[0].message}
     </ModelDetailsLink>
   );
-  const list = messages.slice(0, 5).map((message) => {
+  const list: ListItem[] = messages.slice(0, 5).map((message) => {
```
Comment (huwshimi) [inline src/components/ModelTableList/StatusGroup.tsx:58]: This was just drive-by fix as this type was updated in React Components.

File: src/components/PanelHeader/PanelHeader.tsx [hunk 2]
```diff
@@ -14,7 +14,7 @@ export default function PanelHeader({
   return (
     <div className="p-panel__header">
       <div className="p-panel__title" id={id}>
-        {title}
+        <h5 className="u-no-margin">{title}</h5>
```
Comment (huwshimi) [inline src/components/PanelHeader/PanelHeader.tsx:17]: I've made all the panel headers look the same.

File: src/components/Panel/Panel.tsx [hunk 3]
```diff
@@ -8,14 +8,18 @@ import Aside from "components/Aside/Aside";
...
 import PanelHeader from "components/PanelHeader/PanelHeader";
 
+import "./_panel.scss";
+
 type Props = PropsWithSpread<
   {
     checkCanClose?: (e: KeyboardEvent | MouseEvent) => boolean;
-    panelClassName: string;
+    drawer?: ReactNode;
+    panelClassName?: string;
```
Comment (huwshimi) [inline src/components/Panel/Panel.tsx:17]: I've made this optional now that each panel doesn't require its own styling.

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 4]
```diff
@@ -209,54 +207,45 @@ export default function ActionsPanel(): JSX.Element {
...
+          Run action
+        </Button>
+      }
       width="narrow"
-      panelClassName="actions-panel"
       data-testid={TestId.PANEL}
       title={generateTitle()}
       onRemovePanelQueryParams={handleRemovePanelQueryParams}
     >
-      <>
```
Comment (huwshimi) [inline src/panels/ActionsPanel/ActionsPanel.tsx:218]: This is mostly indentation changes except where I've moved the button to the draw prop.

File: src/panels/ActionsPanel/CharmActionsPanel.tsx [hunk 5]
```diff
@@ -177,49 +171,43 @@ export default function CharmActionsPanel({
...
+        </Button>
+      }
       width="narrow"
-      panelClassName={ClassName.PANEL}
       data-testid={TestId.PANEL}
       title={<CharmActionsPanelTitle charmURL={charmURL} />}
       onRemovePanelQueryParams={onRemovePanelQueryParams}
       initial={false}
     >
-      <>
```
Comment (huwshimi) [inline src/panels/ActionsPanel/CharmActionsPanel.tsx:187]: More indentation changes except for moving the buttons to the drawer.

File: src/panels/CharmsAndActionsPanel/CharmActionsPanelTitle/CharmActionsPanelTitle.test.tsx [hunk 6]
```diff
@@ -86,7 +86,6 @@ describe("CharmActionsPanelTitle", () => {
 
     const title = screen.getByText("1 application (2 units) selected");
     expect(title).toBeVisible();
-    expect(title.tagName).toBe("H5");
```
Comment (huwshimi) [inline src/panels/CharmsAndActionsPanel/CharmActionsPanelTitle/CharmActionsPanelTitle.test.tsx:89]: This element doesn't exist in this component as it was moved to the panel header.

File: src/panels/CharmsPanel/CharmsPanel.tsx [hunk 7]
```diff
@@ -38,41 +32,32 @@ export default function CharmsPanel({
...
+        </Button>
+      }
       width="narrow"
-      panelClassName={ClassName.PANEL}
       data-testid={TestId.PANEL}
       title={Label.PANEL_TITLE}
       onRemovePanelQueryParams={onRemovePanelQueryParams}
       loading={isLoading}
     >
-      <div className="p-panel__content p-panel_content--padded">
```
Comment (huwshimi) [inline src/panels/CharmsPanel/CharmsPanel.tsx:48]: Again, mostly just indentation changes except for moving the buttons to the drawer.

File: src/panels/ConfigPanel/BooleanConfig.tsx [hunk 8]
```diff
@@ -25,8 +25,8 @@ export default function BooleanConfig({
       setSelectedConfig={setSelectedConfig}
       setNewValue={setNewValue}
       input={(value) => (
-        <div className="row u-no-padding">
```
Comment (huwshimi) [inline src/panels/ConfigPanel/BooleanConfig.tsx:28]: This was breaking the layout at some screen sizes.

File: src/panels/ConfigPanel/ConfigPanel.tsx [hunk 9]
```diff
@@ -268,110 +269,100 @@ export default function ConfigPanel(): JSX.Element {
...
+              <div className="u-align--center">
+                <NoDescriptionMessage />
+              </div>
+            )}
+          </div>
+        ) : null
+      }
     >
       <>
-        {!isLoading && (!config || Object.keys(config).length === 0) ? (
```
Comment (huwshimi) [inline src/panels/ConfigPanel/ConfigPanel.tsx:286]: This one has a few changes. The split wrappers were moved to the panel component. The right hand side content was moved to the `splitContent` prop and the buttons to the drawer.

File: src/panels/ConfigPanel/_config-panel.scss [hunk 10]
```diff
@@ -267,8 +186,6 @@ $panel-padding: 1.5rem;
   }
 
   textarea {
-    border: 1px solid $color-grey60;
-    border-radius: 2px;
```
Comment (huwshimi) [inline src/panels/ConfigPanel/_config-panel.scss:271]: This was a fix to make the textareas match the new Vanilla styling.

File: src/panels/RegisterController/RegisterController.tsx [hunk 11]
```diff
@@ -85,187 +85,187 @@ export default function RegisterController() {
...
+            </Button>
+          </Col>
+        </Row>
+      }
       panelClassName="register-controller"
-      title={<h4>{Label.TITLE}</h4>}
+      title={Label.TITLE}
       onRemovePanelQueryParams={handleRemovePanelQueryParams}
     >
-      <div className="p-panel__content p-panel_content--padded">
```
Comment (huwshimi) [inline src/panels/RegisterController/RegisterController.tsx:92]: More indentation changes and moving the buttons.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 12]
```diff
@@ -282,70 +282,35 @@ export default function ShareModel() {
...
-          ) : (
-            <div className="title-wrapper__heading">
-              <h5>
-                <i className="p-icon--share is-inline"></i> Model access:{" "}
-                {modelName}
-              </h5>
-            </div>
-          )}
-        </div>
+        showAddNewUser ? (
```
Comment (huwshimi) [inline src/panels/ShareModelPanel/ShareModel.tsx:290]: Mostly indentation changes from moving the `<h5>` to the panel header.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 13]
```diff
@@ -282,70 +282,35 @@ export default function ShareModel() {
...
-                access={usersAccess?.[username]}
-                isOwner={isOwner(username)}
-                removeUser={handleRemoveUser}
-                accessSelectChange={handleAccessSelectChange}
-              />
-            );
-          })}
-        </div>
-        <div className="aside-split-col add-new-user">
+      splitContent={
```
Comment (huwshimi) [inline src/panels/ShareModelPanel/ShareModel.tsx:312]: The right hand side content was moved here.

**Review by vladimir-cucu — APPROVED** [review]
Looks good! The only small remark I have is that when the panel is split, then we would have that the entire panel content is scalable vertically. I assume we would want to have the individual sides of the panel to be scrollable independently instead?

**Comment by huwshimi** [comment]
> Looks good! The only small remark I have is that when the panel is split, then we would have that the entire panel content is scalable vertically. I assume we would want to have the individual sides of the panel to be scrollable independently instead?

I think this is behaving as expected (assuming I'm understanding your question correctly). The content on the right hand side becomes sticky as you scroll down (in our current panels the right side has additional content related to the left hand side).

Like this:

Screen Recording 2023-08-10 at 10 33 26 am

At the moment all the panels are supposed to behave like this, but if we need to add a panel with independent left/right scrolling we could add a modifier to the panel to not make the right hand side sticky and instead set it to scroll like the left side does (potentially applying the same `side-panel__content-scrolling` class).

I'm going to merge this to unblock the filter PR, but let's talk about this later to make sure I haven't missed something!
